controllers/flinkcluster_reconciler.go

```diff
@@ -520,6 +534,39 @@ func (reconciler *ClusterReconciler) cancelFlinkJob(jobID string, takeSavepoint
 	return reconciler.flinkClient.StopJob(apiBaseURL, jobID)
 }
 
+// Trigger savepoint if it is possible, then return the savepoint status to update.
+// If triggering savepoint is skipped or triggered savepoint was created, proceed to stop the job.
+func (reconciler *ClusterReconciler) cancelFlinkJobAsync(jobID string, takeSavepoint bool) (newSavepointStatus *v1beta1.SavepointStatus, err error) {
+	var log = reconciler.log
+	var cluster = reconciler.observed.cluster
+	var observedSavepoint = reconciler.observed.cluster.Status.Savepoint
+	if takeSavepoint && reconciler.canTakeSavepoint() {
+		if observedSavepoint != nil && observedSavepoint.TriggerID == "" {
+			newSavepointStatus, err = reconciler.takeSavepointAsync(jobID, SavepointTriggerReasonJobCancel)
+			if err != nil {
+				log.Info("Failed to trigger savepoint.")
+				return newSavepointStatus, fmt.Errorf("failed to trigger savepoint: %v", err)
+			}
+			log.Info("Triggered savepoint and wait it is completed.")
+			return newSavepointStatus, nil
+		}
+		if observedSavepoint != nil && observedSavepoint.State == SavepointStateSucceeded {
+			log.Info("Successfully savepoint created. Continue to stop job.")
```

Comment:
I feel it is a little bit confusing to put this logic in this function `cancelFlinkJobAsync `.

Reply:
You're right. I fixed.